brush-core/src/builtins/mapfile.rs

```diff
@@ -66,14 +61,20 @@ impl builtins::Command for MapFileCommand {
         // Read!
         let results = self.read_entries(input_file)?;
 
-        // Assign!
-        context.shell.env.update_or_add(
-            &self.array_var_name,
-            variables::ShellValueLiteral::Array(results),
-            |_| Ok(()),
-            env::EnvironmentLookup::Anywhere,
-            env::EnvironmentScope::Global,
-        )?;
+        for (elem_idx, result) in results.0.into_iter().enumerate() {
+            // If the user is getting to wraparounds in *bash*, they got bigger problems.
+            #[allow(clippy::cast_possible_wrap)]
+            let elem_idx = elem_idx as i64;
+            // Assign!
+            context.shell.env.update_or_add_array_element(
```

Comment:
Thanks for tackling this option. I think we're going to need a wide range of compat tests in `mapfile.yaml`, run both with `-O` and without it--and with different values passed to `-O`, e.g.:

* Variable doesn't already exist before call.
* Variable exists but is a non-array, plain string variable.
* Variable exists and is a regular indexed array.
* Variable exists and is an *associative* array.
* `-O` is passed a negative value

I'm sure there may be other interesting cases, but without the tests it's hard to be certain whether the implementation of `.update_or_add_array_element()` will match the required behavior here--particularly with respect to variable conversion.